subsys/llext/llext.c

```diff
@@ -709,48 +909,67 @@ int llext_load(struct llext_loader *ldr, const char *name, struct llext **ext)
 		}
 
 		ldr->hdr = ehdr;
-		ret = do_llext_load(ldr, *ext);
-		break;
-	default:
-		LOG_ERR("Unsupported elf file type %x", ehdr.e_type);
-		*ext = NULL;
-		return -EINVAL;
-	}
+		ret = do_llext_load(ldr, *ext, do_local);
+		if (ret < 0) {
+			k_heap_free(&llext_heap, *ext);
+			*ext = NULL;
+			return ret;
+		}
 
-	if (ret == 0) {
 		strncpy((*ext)->name, name, sizeof((*ext)->name));
 		(*ext)->name[sizeof((*ext)->name) - 1] = '\0';
+		(*ext)->use_count++;
+
 		sys_slist_append(&_llext_list, &(*ext)->_llext_list);
 		LOG_INF("Loaded extension %s", (*ext)->name);
+
+		break;
+	default:
+		LOG_ERR("Unsupported elf file type %x", ehdr.e_type);
+		return -EINVAL;
 	}
 
-	return ret;
+	return 0;
 }
 
-void llext_unload(struct llext *ext)
+void llext_unload(struct llext **ext, void (*free)(struct llext *, const void *), const void *ctx)
 {
-	__ASSERT(ext, "Expected non-null extension");
+	__ASSERT(*ext, "Expected non-null extension");
+	struct llext *tmp = *ext;
 
-	sys_slist_find_and_remove(&_llext_list, &ext->_llext_list);
+	k_mutex_lock(&llext_lock, K_FOREVER);
+	if (tmp->use_count-- != 1) {
+		k_mutex_unlock(&llext_lock);
+		return;
+	}
+
+	if (free) {
+		free(tmp, ctx);
+	}
+
+	/* FIXME: protect the global list */
```

Comment:
Looks like its in a mutex now, so is protected

Reply:
only here, but it's still unprotected in most locations.

Reply:
@teburd please check the update, I added protection to the list too